lib/ash_authentication/add_ons/log_out_everywhere/action.ex

```diff
@@ -0,0 +1,80 @@
+defmodule AshAuthentication.AddOn.LogOutEverywhere.Action do
+  @moduledoc """
+  Revokes all tokens for the specified user.
+  """
+  use Ash.Resource.Actions.Implementation
+
+  alias Ash.{
+    ActionInput,
+    Error.Framework.AssumptionFailed,
+    Query,
+    Resource
+  }
+
+  alias AshAuthentication.Info
+
+  @doc false
+  @impl true
+  def run(input, _opts, _context) do
+    case Info.strategy_for_action(input.resource, input.action.name) do
+      {:ok, strategy} ->
+        really_run_action(input, strategy)
+
+      :error ->
+        raise AssumptionFailed,
+          message: "Action does not correlate with an authentication strategy"
+    end
+  end
+
+  defp really_run_action(input, strategy) do
+    with {:ok, user_id} <- ActionInput.fetch_argument(input, strategy.argument_name),
+         {:ok, user} <- get_user_by_id(user_id, strategy) do
+      subject = AshAuthentication.user_to_subject(user)
+      revoke_all_tokens_for_subject(subject, strategy)
+    end
+  end
+
+  defp get_user_by_id(user_id, strategy) do
+    with [id] <- Resource.Info.primary_key(strategy.resource) do
+      strategy.resource
+      |> Query.new()
+      |> Query.set_context(%{private: %{ash_authentication?: true}})
+      |> Query.do_filter([{id, user_id}])
+      |> Ash.read_one()
+    end
+  end
+
+  defp revoke_all_tokens_for_subject(subject, strategy) do
+    with {:ok, token_resource} <- Info.authentication_tokens_token_resource(strategy.resource),
+         {:ok, tokens} <- get_all_tokens_for_subject(token_resource, subject) do
+      tokens
+      |> Stream.map(
+        &%{
+          jti: &1.jti,
+          purpose: "revocation",
```

Comment:
We explicitly add new tokens to revoke the old ones, we wouldn't simply delete all the existing tokens? That's what I was expecting to see here

Reply:
Good point. Seems like we should only be revoking specific kinds of tokens? Or maybe the purposes revoked should be configurable and should default to the ones we care about like sign_in, magic_link, password_reset etc?

Reply:
As for revoking old tokens, I also think that would be a better default strategy, but the default setup does not `require_token_presence_for_authentication`. @jimsynz perhaps we should modify token revocation to just delete tokens when `require_token_presence_for_authentication`? Then we also don't need to check for revocation tokens if that setting is true also. Users can still opt out of the requirement to store all tokens, but when they opt in, AA will then make fewer queries and behave a bit simpler (the way @sevenseacat originally expected it to).

Reply:
we upsert them which really since we queried for them in the first place is just an update. maybe I should just make it an update.

Reply:
From what we were seeing, we do not upsert them. There is not an identity on the jit.

Reply:
The JTI is the primary key

Reply:
Oh, where are we configuring it to be an upsert when revoking?

Reply:
aw damnit...its the generators again, they are adding a uuid primary key 😢.

So we need another warning and fix to that. Its fine because current logic checks for a revocation token 😂

Reply:
Fix is here, and users will need to follow the instructions when upgrading:

https://github.com/team-alembic/ash_authentication/pull/908

Reply:
Alright, so once that PR is merged, we can merge this one, and then we can, as a follow up, remove any place that we are looking up revocation tokens if `require_token_presence_for_authentication?` is set to `true`. Finally, we will want to modify the installers to make all of this the default behavior. We can do that in a follow up to this PR.

Reply:
Well, not in that order. We should modify the installers first, because the avoiding a lookup thing is just a perf improvement.

Reply:
Okay, this now does a bulk update.